### ISSUE TITLE
Slow migration from 2.4 to 3.0

### ISSUE THREAD

**Issue by brchristian** [spree/spree#6726]
I'm updating my app from `2-4-stable` to `3-0-stable` and one of the migrations for our app requires 24 hours(!) to complete.

Here's what the logs look like:

    Spree::Payment Exists (79.3ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P4KSFWXE' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P4KSFWXE' WHERE `spree_payments`.`id` = 9938
    Spree::Payment Exists (79.0ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PAC1C6ND' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PAC1C6ND' WHERE `spree_payments`.`id` = 9939
    Spree::Payment Exists (82.0ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P4WID31Y' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P4WID31Y' WHERE `spree_payments`.`id` = 9940
    Spree::Payment Exists (83.1ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PS3KRTAD' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PS3KRTAD' WHERE `spree_payments`.`id` = 9941
    Spree::Payment Exists (100.3ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P5NDHM5A' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P5NDHM5A' WHERE `spree_payments`.`id` = 9942
    Spree::Payment Exists (97.3ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PJTV1VBB' LIMIT 1
    SQL (1.2ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PJTV1VBB' WHERE `spree_payments`.`id` = 9943
    Spree::Payment Exists (96.3ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PWLOTU68' LIMIT 1
    SQL (1.3ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PWLOTU68' WHERE `spree_payments`.`id` = 9944
    Spree::Payment Exists (92.7ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PGFCA4TU' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PGFCA4TU' WHERE `spree_payments`.`id` = 9945
    Spree::Payment Exists (88.6ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PH2KHKF8' LIMIT 1
    SQL (1.4ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PH2KHKF8' WHERE `spree_payments`.`id` = 9946
    Spree::Payment Exists (84.8ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P7TKN73B' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P7TKN73B' WHERE `spree_payments`.`id` = 9947
    Spree::Payment Exists (92.5ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P0CUJ7R5' LIMIT 1
    SQL (0.9ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P0CUJ7R5' WHERE `spree_payments`.`id` = 9948
    Spree::Payment Exists (88.1ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PK58QYKA' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PK58QYKA' WHERE `spree_payments`.`id` = 9949
    Spree::Payment Exists (89.6ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P2O61FY9' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P2O61FY9' WHERE `spree_payments`.`id` = 9950
    Spree::Payment Exists (95.9ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PEG7SBWH' LIMIT 1
    SQL (1.2ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PEG7SBWH' WHERE `spree_payments`.`id` = 9951
    Spree::Payment Exists (87.9ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PL51GLRE' LIMIT 1
    SQL (1.3ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PL51GLRE' WHERE `spree_payments`.`id` = 9952
    Spree::Payment Exists (92.3ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PEQYQLVE' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PEQYQLVE' WHERE `spree_payments`.`id` = 9953
    Spree::Payment Exists (90.8ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PAR55K43' LIMIT 1
    SQL (1.1ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PAR55K43' WHERE `spree_payments`.`id` = 9954
    Spree::Payment Exists (88.1ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PMTMO6EO' LIMIT 1
    SQL (0.7ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PMTMO6EO' WHERE `spree_payments`.`id` = 9955
    Spree::Payment Exists (90.2ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PTHY7MI1' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PTHY7MI1' WHERE `spree_payments`.`id` = 9956
    Spree::Payment Exists (87.1ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'PMYDWEZX' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'PMYDWEZX' WHERE `spree_payments`.`id` = 9957
    Spree::Payment Exists (80.8ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P1P38QAP' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P1P38QAP' WHERE `spree_payments`.`id` = 9958
    Spree::Payment Exists (85.4ms)  SELECT  1 AS one FROM `spree_payments` WHERE `spree_payments`.`number` = 'P1MJI75T' LIMIT 1
    SQL (0.6ms)  UPDATE `spree_payments` SET `spree_payments`.`number` = 'P1MJI75T' WHERE `spree_payments`.`id` = 9959

The problem seems to be the `exists` call here: https://github.com/spree/spree/blob/3-0-stable/core/app/models/concerns/spree/number_generator.rb#L30

In our case, we've got something on the order of a million payments -- is there a more efficient way to do this in bulk without making a million `exists` calls at 85ms apiece? Ideas welcome!

**Comment by brchristian** [comment]
I've found that using a (temporary) db index will massively cut down the migration time -- I've created a PR at https://github.com/spree/spree/pull/6727 that will help other users in a similar situation.

**Comment by jasonfb** [comment]
@brchristian -- I invented this Gem to deal with this problem. It does not reduce the time of the migration-- but  it allows you to run data migrations while your app is online, so you don't have to take the app offline for 24 hours.

https://github.com/jasonfb/nondestructive_migrations

Running huge migrations in DB migrations is an anti-pattern in my opinion. This gem eases the pain points for almost all long-running non-schema changing migrations

**Comment by JDutil** [comment]
Closing in favor of PR

**Comment by brchristian** [comment]
:+1:

**Comment by brchristian** [comment]
@jasonfb Thanks for that, looks really useful!

@JDutil FWIW I have found myself tweaking some of the Spree migrations to run while the app is live in the past, as @jasonfb is saying. In these cases, I've found it really helpful to add a `.reverse` to the Rails scopes, so that for instance new or recent orders get migrated immediately and the migration works its way back into the archival stuff.

It might not necessarily be worth establishing it as a Spree convention to add a `.reverse` to all migration scopes -- e.g., `Spree::Payment.where(number: nil).reverse.find_each` -- but I thought I'd mention it in this thread, in case it's helpful feedback for the Spree team, or helpful to other app owners in similar situations. It's enabled us to run migrations while the app is live, and be assured that active orders are processed in the first few seconds, even if the old orders from years ago take a few minutes or hours to migrate.
